tap_redshift/__init__.py

```diff
@@ -0,0 +1,421 @@
+                            c['name']: schema_for_column(c) for c in cols})
+        tap_stream_id = '{}-{}'.format(dbname, table_name)
+        entry = CatalogEntry(
+                    database=dbname,
+                    tap_stream_id=tap_stream_id,
+                    stream=table_name,
+                    schema=schema,
+                    table=table_name)
+        table_type = [t for (t) in table_spec]
+        entry.is_view = table_type == 'VIEW'
+        entries.append(entry)
+
+    return Catalog(entries)
+
+
+def do_discover():
+    discover_catalog().dump()
+
+
+def schema_for_column(c):
+    '''Returns the Schema object for the given Column.'''
+    column_type = c['type'].lower()
+    inclusion = 'available'
+
+    result = Schema(inclusion=inclusion)
+
+    if column_type == 'bool':
+        result.type = ['null', 'boolean']
+
+    elif column_type in BYTES_FOR_INTEGER_TYPE:
+        result.type = ['null', 'integer']
+        bits = BYTES_FOR_INTEGER_TYPE[column_type] * 8
+        result.minimum = 0 - 2 ** (bits - 1)
+        result.maximum = 2 ** (bits - 1) - 1
+
+    elif column_type in FLOAT_TYPES:
+        result.type = ['null', 'number']
+
+    elif column_type == 'decimal':
+        result.type = ['null', 'number']
+        result.exclusiveMaximum = True
+
+    elif column_type in STRING_TYPES:
+        result.type = ['null', 'string']
+
+    elif column_type in DATETIME_TYPES:
+        result.type = ['null', 'string']
+        result.format = 'date-time'
+
+    else:
+        result = Schema(None,
+                        inclusion='unsupported',
+                        description='Unsupported column type {}'
+                        .format(column_type))
+
+    return result
+
+
+def open_connection(**kwargs):
+
+    # For testing purpose
+    if kwargs:
+        args = kwargs.get('mock')
+        host = args['host'],
+        port = args['port'],
+        dbname = args['dbname'],
+        user = args['user'],
+        password = args['password']
+    else:
+        args = utils.parse_args(REQUIRED_CONFIG_KEYS)
+        config = args.config
+        host = config['host'],
+        port = config['port'],
+        dbname = config['dbname'],
+        user = config['user'],
+        password = config['password']
+
+    connection = psycopg2.connect(
+            host=host[0],
+            port=port[0],
+            dbname=dbname[0],
+            user=user[0],
+            password=password)
+    return connection
+
+
+def select_all(query, **kwargs):
+    conn = open_connection(**kwargs)
+    cur = conn.cursor()
+    cur.execute(query)
+    column_specs = cur.fetchall()
+    cur.close()
+    return column_specs
+
+
+def get_stream_version(tap_stream_id, state):
+    return singer.get_bookmark(state,
+                               tap_stream_id,
+                               "version") or int(time.time() * 1000)
+
+
+def row_to_record(catalog_entry, version, row, columns, time_extracted):
```

Comment:
These type conversions seem to have been copied from some other tap implementation. Do they make sense for RedShift? Which additional conversions might be necessary, if any?

Reply:
Looking through RedShift, I'm not sure if any other conversion is needed and rectified this appropriately.